core/services/helper/main.py

```diff
@@ -121,94 +140,191 @@ class SpeedTestResult(BaseModel):
-                        api = response.json()
-                        # Check all available versions
-                        ## The ones that provide a swagger-ui
-                        for path in api["paths"].keys():
-                            with requests.get(f"http://127.0.0.1:{port}{path}", timeout=0.2) as response:
-                                if "swagger-ui" in response.text:
-                                    info.versions += [path]
+                for api_path in Helper.API_CANDIDATE_URLS:
+                    conn = http.client.HTTPConnection("127.0.0.1", port, timeout=0.2)
+                    conn.request("GET", api_path)
+                    response = conn.getresponse()
+                    if response.status != http.client.OK or not Helper._is_json_response(response):
+                        continue
+                    api = json.loads(response.read().decode())
+                    # Check all available versions
+                    ## The ones that provide a swagger-ui
+                    for path in api["paths"].keys():
+                        conn = http.client.HTTPConnection("127.0.0.1", port, timeout=0.2)
+                        conn.request("GET", path)
+                        response = conn.getresponse()
+                        if "swagger-ui" in response.read().decode():
+                            info.versions += [path]
                 break
 
-            except Exception:
+            except http.client.BadStatusLine:
+                pass
+
+            except Exception as e:
+                logging.error(e, exc_info=True)
                 # This should be avoided by the first try block, but better safe than sorry
                 break
+
+            finally:
+                if conn:
+                    conn.close()
+
+        logging.debug(f"{log_msg}: valid. Took: {time.process_time() - start_time} s")
         return info
 
     @staticmethod
     @temporary_cache(timeout_seconds=10)
     def scan_ports() -> List[ServiceInfo]:
+        log_msg = "Running scan_ports"
+        start_time = time.process_time()
+
         # Filter for TCP ports that are listen and can be accessed by external users (server in 0.0.0.0)
-        connections = (
+        connections = [
             connection
             for connection in psutil.net_connections("tcp")
             if connection.status == psutil.CONN_LISTEN and connection.laddr.ip in Helper.LOCALSERVER_CANDIDATES
-        )
+        ]
 
         # And check if there is a webpage available that is not us
         # Use it as a set to remove duplicated ports
         ports = set(connection.laddr.port for connection in connections)
-        services = [Helper.detect_service(port) for port in ports if port != PORT]
-        return [service for service in services if service.valid]
+        services = [
+            Helper.detect_service(port) for port in ports if port != Helper.PORT and port not in Helper.SKIP_PORTS
+        ]
```

Comment:
We could experiment with doing this using the concurrent future executor here as well, which might shape the CPU usage as a higher but shorter peak, reducing the overall response time.